tests/query_params_test.go

```diff
@@ -0,0 +1,216 @@
+	req := []byte("GET /?vegetable=potato&Vegetable=tomato HTTP/1.1\r\n" +
+		"Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		queryParams := req.URL.Query()
+
+		if len(queryParams) != 2 {
+			t.Errorf("len(queryParams): got %d, want %d", len(queryParams), 2)
+		}
+
+		if len(queryParams["vegetable"]) != 1 || len(queryParams["Vegetable"]) != 1 {
+			t.Errorf("Expected one value for query parameter vegetables and one for Vegetables, got %d and %d", len(queryParams["vegetable"]), len(queryParams["Vegetable"]))
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+}
+
+// TestQueryParametersValidUnicode ensures keys and values that contain non-ASCII characters are parsed correctly
+func TestQueryParametersValidUnicode(t *testing.T) {
+	value := "ăȚâȘî"
+	req := []byte("GET /?vegetable=" + value + " HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		if valueReceived := req.URL.Query()["vegetable"][0]; valueReceived != value {
+			t.Errorf("queryParams values: got %s, want %s", valueReceived, value)
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+	key := "ăȚâȘî"
+	req = []byte("GET /?" + key + "=vegetable HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	resp, err = requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		if listLen := len(req.URL.Query()[key]); listLen != 1 {
+			t.Errorf("len(queryParamsKey): got %d, want 1 value for %s", listLen, key)
+		}
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isOkReq(resp) {
+		t.Errorf("response: want %s, got %s", statusOK, string(resp))
+	}
+
+}
+
+// Tests whether passing invalid Unicode will result  in a 400 Bad Request response
+func TestQueryParametersInvalidUnicodes(t *testing.T) {
+	key := "\x0F"
+	req := []byte("GET /?" + key + "=tomato&Vegetable=potato HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
+	resp, err := requesttesting.MakeRequest(context.Background(), req, func(req *http.Request) {
+		t.Error("MakeRequest(): Expected handler not to be called.")
+	})
+	if err != nil {
+		t.Errorf("MakeRequest(): got err %v, want nil", err)
+	}
+	if !isBadReq(resp) {
+		t.Errorf("response: want %s, got %s", statusBadReq, string(resp))
+	}
+
+}
+
+// Verifies behaviour of query parameter parser when passing malformed keys or
+// values, by breaking URL percent-encoding. Percent-encoding is used to encode
+// special characters in URL
+func TestQueryParametersBreakUrlEncoding(t *testing.T) {
+	brokenKeyReq := []byte("GET /?vegetable%=tomato HTTP/1.1\r\n" + "Host: localhost:8080\r\n" + "\r\n")
```

Comment:
This could probably be folded into the table-driven test as well.

Reply:
One of the checks inspects whether breaking the url encoding in the query param key fails and the other inspects the same thing for the query param value. Is it possible to do it?